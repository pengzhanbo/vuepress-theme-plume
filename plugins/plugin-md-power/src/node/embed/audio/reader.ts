import type { PluginWithOptions } from 'markdown-it'
import type { RuleInline } from 'markdown-it/lib/parser_inline.mjs'
import { resolveAttrs } from '../../utils/resolveAttrs.js'

const audioReader: RuleInline = (state, silent) => {
  const max = state.posMax
  let start = state.pos
  let href = ''

  if (state.src.slice(start, start + 13) !== '@[audioReader')
    return false

  // @[audioReader]()
  if (max - start < 17)
    return false

  const labelStart = state.pos + 13
  const labelEnd = state.md.helpers.parseLinkLabel(state, state.pos + 1, true)

  // not found ']'
  if (labelEnd < 0)
    return false

  let pos = labelEnd + 1
  if (pos < max && state.src.charCodeAt(pos) === 0x28 /* ( */) {
    pos++
    const code = state.src.charCodeAt(pos)
    if (code === 0x0A /* \n */ || code === 0x20 /* space */)
      return false

    start = pos
    const res = state.md.helpers.parseLinkDestination(state.src, pos, state.posMax)
    if (res.ok) {
      href = state.md.normalizeLink(res.str)
      if (state.md.validateLink(href)) {
        pos = res.pos
      }
      else {
        href = ''
      }
    }
    if (pos >= max || state.src.charCodeAt(pos) !== 0x29 /* ) */) {
      return false
    }
  }
  else {
    return false
  }

  /* istanbul ignore else -- @preserve */
  if (!silent) {
    state.pos = labelStart
    state.posMax = labelEnd
    const info = state.src.slice(labelStart, labelEnd).trim()
    const { attrs } = resolveAttrs(info)

    const tokenOpen = state.push('audio_reader_open', 'AudioReader', 1)
    tokenOpen.info = info
    tokenOpen.attrs = [['src', href]]

    if (attrs.startTime)
      tokenOpen.attrs.push([':start-time', attrs.startTime])

    if (attrs.endTime)
      tokenOpen.attrs.push([':end-time', attrs.endTime])

    if (attrs.type)
      tokenOpen.attrs.push(['type', attrs.type])

    if (attrs.volume)
      tokenOpen.attrs.push([':volume', attrs.volume])

    if (attrs.title)
      state.push('text', '', 0).content = attrs.title

    state.push('audio_reader_close', 'AudioReader', -1)
  }

  state.pos = pos + 1
  state.posMax = max
  return true
}

export const audioReaderPlugin: PluginWithOptions<never> = md =>
  md.inline.ruler.before('link', 'audio-reader', audioReader)
