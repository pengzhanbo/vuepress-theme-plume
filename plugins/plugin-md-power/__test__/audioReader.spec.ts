import MarkdownIt from 'markdown-it'
import { describe, expect, it } from 'vitest'
import { audioReaderPlugin } from '../src/node/embed/audio/reader.js'

function createMarkdown() {
  return MarkdownIt().use((md) => {
    md.block.ruler.before('code', 'import_code', () => false)
    md.renderer.rules.import_code = () => ''
  }).use(audioReaderPlugin)
}

describe('artPlayerPlugin', () => {
  it('should work', () => {
    const md = createMarkdown()
    const code = `\
@[audioReader](/xxx.mp3) @[audioReader](/xxx.mp3)

@[audioReader autoplay title="title"](/xxx.mp3)

@[audioReader autoplay start-time="0" end-time="99" volume="0.55"](/xxx.mp3)

xxx @[audioReader type="audio/mp3"](/xxx.mp3) xxx
`

    expect(md.render(code)).toMatchSnapshot()
  })

  it('should not work', () => {
    const md = createMarkdown()
    const code = `\
@[audioReader @[audioReader]()

@[audioReader]xxx

@[ audioReader](123456

@[audioReader]((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((()

@[ audioReader](/xxx.mp3) xxx

@[audioReader]( xxx.mp3) xxx

@[audioReader](javascript:alert(1)) xxx
`

    expect(md.render(code)).toMatchSnapshot()
  })
})
