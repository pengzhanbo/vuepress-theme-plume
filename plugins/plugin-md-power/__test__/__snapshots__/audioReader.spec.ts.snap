// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`artPlayerPlugin > should not work 1`] = `
"<p>@[audioReader @<a href="">audioReader</a></p>
<p>@[audioReader]xxx</p>
<p>@[ audioReader](123456</p>
<p>@[audioReader]((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((()</p>
<p>@<a href="/xxx.mp3"> audioReader</a> xxx</p>
<p>@<a href="xxx.mp3">audioReader</a> xxx</p>
<p>@[audioReader](javascript:alert(1)) xxx</p>
"
`;

exports[`artPlayerPlugin > should work 1`] = `
"<p><AudioReader src="/xxx.mp3"></AudioReader> <AudioReader src="/xxx.mp3"></AudioReader></p>
<p><AudioReader src="/xxx.mp3">title</AudioReader></p>
<p><AudioReader src="/xxx.mp3" :start-time="0" :end-time="99" :volume="0.55"></AudioReader></p>
<p>xxx <AudioReader src="/xxx.mp3" type="audio/mp3"></AudioReader> xxx</p>
"
`;
